packages/babel-parser/src/plugins/flow.js

```diff
@@ -2679,23 +2679,44 @@ export default (superClass: Class<Parser>): Class<Parser> =>
 
           const arrowExpression = this.forwardNoArrowParamsConversionAt(
             typeParameters,
-            () =>
-              super.parseMaybeAssign(
+            () => {
+              const result = super.parseMaybeAssign(
                 refExpressionErrors,
                 afterLeftParse,
                 refNeedsArrowPos,
-              ),
+              );
+
+              this.resetStartLocationFromNode(result, typeParameters);
+
+              return result;
+            },
           );
-          arrowExpression.typeParameters = typeParameters;
-          this.resetStartLocationFromNode(arrowExpression, typeParameters);
+
+          // The above can return a TypeCastExpression when the arrow
+          // expression is not wrapped in parens.
+          const expr = this.maybeUnwrapTypeCastExpression(arrowExpression);
```

Comment:
Can you add a new test case?
```js
<T>(() => {}: any);
```
should throw.

Reply:
Nice catch!

I'll take a look at the two other issues you found tomorrow.

Reply:
Yeah. Those two are not blockers, we can address those two in a separate PR.